### DISCUSSION TITLE
IPSec tunnel/SA re-establishment issue

### DISCUSSION THREAD

**Discussion by lkpatel** [strongswan/strongswan#2262] [Q&A]
Hi,
I am new to strongswan. I have a use case where two peers are having two parallel connections from different ISP.  When all links are up and both tunnels/SAs are up, traffic goes from one tunnel and if one link goes down, the traffic falls back to other link as expected. 
Now if other remaining link also goes down, the traffic stop as expected. Issue is that if suppose any of the link comes back alive, the tunnel/SA is not being re-established. Please see the attached diagram for my setup. Any help would be appreciated.
My config are as per following-

_connections {
    gw-gw1 {
      remote_addrs = sun
      local_addrs = moon
      local {
        auth = psk
        id = moon
      }
      remote {
        auth = psk
        id = sun
      }
      children {
        net-net1 {
          local_ts  = 10.10.11.0/24
          remote_ts = 10.10.10.0/24
          start_action = trap|start
          esp_proposals = aes128-sha256-modp3072
        }
      }
      unique = never
      mobike = no
      version = 2
      proposals = aes128-sha256-modp3072
      dpd_timeout = 1
      dpd_delay = 1
      reauth_time = 60
    }
gw-gw2 {
      remote_addrs = sun
      local_addrs = moon
      local {
        auth = psk
        id = moon
      }
      remote {
        auth = psk
        id = sun
      }
      children {
        net-net2 {
          local_ts  = 10.10.11.0/24
          remote_ts = 10.10.10.0/24
          start_action = trap|start
          esp_proposals = aes128-sha256-modp3072
        }
      }
      unique = never
      mobike = no
      version = 2
      proposals = aes128-sha256-modp3072
      dpd_timeout = 1
      dpd_delay = 1
      reauth_time = 60
   }
}_

![Parallel Connection Setup](https://github.com/strongswan/strongswan/assets/2733730/9a0b9ef2-4d2e-4325-a742-aa730a084ad0)

**Comment by tobiasbrunner** [comment]
> When all links are up and both tunnels/SAs are up

Not sure what you mean because your config will only result in two SAs if both peers initiate them concurrently (which is fine, by the way). Adding the same exact config to the file multiple times is completely useless. If you want separate SAs for each link, you'd have to set the local and remote IPs explicitly (on both ends) and disable MOBIKE so there is no automatic switch from one IP to another. However, the latter is exactly for the purpose of updating an existing SA should the current path not be available anymore. For it to work there have to be route/interface changes that the daemon can detect if a link goes down (that might be the case, though, if it worked as expected so far, you can check the log for details).

But note that with a single config and MOBIKE, the daemon will only try the first IP available when initiating an SA from scratch. So if only one link is up, the daemon might try to create the SA over the link that's currently down (round-robin DNS could be an option as the daemon does resolve names anew when retrying to establish an SA, how quickly that happens depends on the [retransmission settings](https://docs.strongswan.org/docs/5.9/config/retransmission.html)).

> Issue is that if suppose any of the link comes back alive, the tunnel/SA is not being re-established.

Read the log. But since you added `trap` to `start action` an SA should get triggered on-demand by traffic matching the traffic selectors (it also avoids traffic leaks if no SA is up). However, if there is no such traffic, the SA will stay down (but is apparently also not required currently) unless you explicitly add `dpd_action=restart` so it gets recreated after a DPD timeout. Note that trap policies might not work well with parallel SAs (i.e. one SA for each link) as only one SA will be triggered by the kernel's acquire so that could be the one over the link that's currently down (and unlike with round-robin DNS, the next attempt after the timeout will trigger the same SA again).

You might also want to consider [route-based](https://docs.strongswan.org/docs/5.9/features/routeBasedVpn.html) approaches.

**Comment by lkpatel** [comment]
@tobiasbrunner thanks for your kind reply. I removed parallel path and now there is only one path between peers. Accordingly, I change the config also. Even then, if links goes down for some time( cable removed to simulate link down scenario) and comes back alive, SA is not reestablishing. I have kept ping traffic on in all cases. What settings are required so that SA are reestablished when links come back alive ?

**Reply by tobiasbrunner** [reply]
Read the logs!

**Reply by lkpatel** [reply]
After link goes down- I get following logs

_**May 29 16:48:52 moon kernel: igb 0000:06:00.0 eno3: igb: eno3 NIC Link is Down**
May 29 16:48:56 moon charon-systemd[252165]: retransmit 1 of request with message ID 10
May 29 16:48:56 moon charon-systemd[252165]: sending packet: from 192.168.0.2[4500] to 192.168.0.1[4500] (80 bytes)
May 29 16:49:03 moon charon-systemd[252165]: retransmit 2 of request with message ID 10
May 29 16:49:03 moon charon-systemd[252165]: sending packet: from 192.168.0.2[4500] to 192.168.0.1[4500] (80 bytes)
May 29 16:49:16 moon charon-systemd[252165]: retransmit 3 of request with message ID 10
May 29 16:49:16 moon charon-systemd[252165]: sending packet: from 192.168.0.2[4500] to 192.168.0.1[4500] (80 bytes)
May 29 16:49:37 moon charon-systemd[252165]: reauthenticating IKE_SA gw-gw1[2]
May 29 16:49:40 moon charon-systemd[252165]: retransmit 4 of request with message ID 10
May 29 16:49:40 moon charon-systemd[252165]: sending packet: from 192.168.0.2[4500] to 192.168.0.1[4500] (80 bytes)
May 29 16:50:22 moon charon-systemd[252165]: retransmit 5 of request with message ID 10
May 29 16:50:22 moon charon-systemd[252165]: sending packet: from 192.168.0.2[4500] to 192.168.0.1[4500] (80 bytes)_
May 29 16:51:37 moon charon-systemd[252165]: giving up after 5 retransmits

and SA get cleared.

_May 29 17:28:30 moon charon-custom[253310]: 11[IKE] giving up after 5 retransmits
May 29 17:28:30 moon charon-custom[253310]: 11[MGR] checkin and destroy IKE_SA gw-gw1[1]
May 29 17:28:30 moon charon-custom[253310]: 11[IKE] IKE_SA gw-gw1[1] state change: ESTABLISHED => DESTROYING
May 29 17:28:30 moon charon-custom[253310]: 11[CHD] CHILD_SA net-net1{2} state change: INSTALLED => DESTROYING
May 29 17:28:30 moon charon-custom[253310]: 11[ESP] deleting policy 10.10.11.0/24 === 10.10.10.0/24 out
May 29 17:28:30 moon charon-custom[253310]: 11[ESP] deleting policy 10.10.10.0/24 === 10.10.11.0/24 in
May 29 17:28:30 moon charon-custom[253310]: 11[ESP] deleted inbound SAD entry with SPI 66ca290c
May 29 17:28:30 moon charon-custom[253310]: 11[ESP] deleted outbound SAD entry with SPI 98ce730c
May 29 17:28:30 moon charon-custom[253310]: 11[MGR] checkin and destroy of IKE_SA successful
May 29 17:28:31 moon charon-custom[253310]: 03[ESP] could not find an outbound IPsec SA for reqid {1}, dropping packet_

After link get restored, I get following logs-

_**May 29 16:56:50 moon kernel: igb 0000:06:00.0 eno3: igb: eno3 NIC Link is Up 1000 Mbps Full Duplex, Flow Control: RX/TX**
May 29 16:56:50 moon NetworkManager[1005]: <info>  [1716982010.7970] device (eno3): carrier: link connected
May 29 16:56:50 moon charon-systemd[252165]: could not find an outbound IPsec SA for reqid {1}, dropping packet
May 29 16:56:51 moon charon-systemd[252165]: could not find an outbound IPsec SA for reqid {1}, dropping packet
May 29 16:56:52 moon charon-systemd[252165]: could not find an outbound IPsec SA for reqid {1}, dropping packet_

It seems, there is no attempt of SA establishment even after link up.

I will try to enhance the log level to find out more information.

Regards

**Reply by tobiasbrunner** [reply]
> May 29 16:56:50 moon charon-systemd[252165]: could not find an outbound IPsec SA for reqid {1}, dropping packet

These messages are from our userland IPsec implementation. So you are apparently using the _kernel-libipsec_ plugin, which is [not generally recommended](https://docs.strongswan.org/docs/5.9/plugins/kernel-libipsec.html) and which also only supports trap policies since 5.9.11 (however, the messages you see show that you use an older version). I suggest you disable that plugin and use the kernel's IPsec implementation via _kernel-netlink_ plugin instead.

**Reply by lkpatel** [reply]
Thanks for your inputs. I am using Rocky Linux 9 and RPM based installation of Strongswan version 5.9.10 . I will try to install the latest possible version and with kernel-netlink plugin and do the test again. 

Regards

**Reply by tobiasbrunner** [reply]
You can just disable the plugin (see [the docs](https://docs.strongswan.org/docs/5.9/plugins/pluginLoad.html)).

**Comment by lkpatel** [comment]
I complied the latest version with --enable-systemd options and it works like a charm. For single connection, everything is working fine. After link re-establishment, SA come back without any issue. 
Now I have established two parallel links where each link have IP in separate subnet and routes for connected networks have been added through both links with different cost in both peers.  

Peer "**moon"** config is as per following-

_connections {
    gw-gw1 {
      remote_addrs = sun1
      local_addrs = moon1
      local {
        auth = psk
        id = moon1
      }
      remote {
        auth = psk
        id = sun1
      }
      children {
        net-net1 {
          local_ts  = 10.10.11.0/24
          remote_ts = 10.10.10.0/24
          start_action = trap|start
          esp_proposals = aes128-sha256-modp3072
          dpd_action=restart
        }
      }
      unique = never
      mobike = no
      version = 2
      proposals = aes128-sha256-modp3072
      dpd_timeout = 1
      dpd_delay = 1
      reauth_time = 60
    }
    gw-gw2 {
      remote_addrs = sun2
      local_addrs = moon2
      local {
        auth = psk
        id = moon2
      }
      remote {
        auth = psk
        id = sun2
      }
      children {
        net-net2 {
          local_ts  = 10.10.11.0/24
          remote_ts = 10.10.10.0/24
          start_action = trap|start
          esp_proposals = aes128-sha256-modp3072
          dpd_action=restart
        }
      }
      unique = never
      mobike = no
      version = 2
      proposals = aes128-sha256-modp3072
      dpd_timeout = 1
      dpd_delay = 1
     reauth_time = 60
   }
}_

opposite config is there in other peer "**sun**".

Now when I start/restart the strongswan service, both SA are established with out any issue. Now I have two SAs
**gw-gw1 : 192.168.0.1 <=> 192.168.0.1
gw-gw2:  192.168.1.1 <=> 192.168.1.1**
Now suppose I make the active (though which traffic is passing) SA  "gw-gw1" down by pulling the cable, traffic falls back to other available SA "gw-gw2". But If now I make "gw-gw2" SA down and try to make "gw-gw-1" SA up, it does not happen. Logs just show that peers are just keep on trying to establish SA over "gw-gw2" ie. 192.168.1.1 <=> 192.168.1.2

 _May 30 17:16:42 sun charon-systemd[394617]: creating delete job for CHILD_SA ESP/0x00000000/192.168.1.2
May 30 17:16:42 sun charon-systemd[394617]: CHILD_SA ESP/0x00000000/192.168.1.2 not found for delete
May 30 17:16:42 sun charon-systemd[394617]: giving up after 5 retransmits
May 30 17:16:42 sun charon-systemd[394617]: establishing IKE_SA failed, peer not responding
May 30 17:16:47 sun charon-systemd[394617]: creating acquire job for policy 10.10.10.10/32[icmp/8] === 10.10.11.10/32[icmp/8] with reqid {1}
May 30 17:16:47 sun charon-systemd[394617]: initiating IKE_SA gw-gw2[5] to 192.168.1.2
May 30 17:16:47 sun charon-systemd[394617]: generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
May 30 17:16:47 sun charon-systemd[394617]: sending packet: from 192.168.1.1[500] to 192.168.1.2[500] (592 bytes)
May 30 17:16:51 sun charon-systemd[394617]: retransmit 1 of request with message ID 0
May 30 17:16:51 sun charon-systemd[394617]: sending packet: from 192.168.1.1[500] to 192.168.1.2[500] (592 bytes)
May 30 17:16:58 sun charon-systemd[394617]: retransmit 2 of request with message ID 0
May 30 17:16:58 sun charon-systemd[394617]: sending packet: from 192.168.1.1[500] to 192.168.1.2[500] (592 bytes)
May 30 17:17:11 sun charon-systemd[394617]: retransmit 3 of request with message ID 0
May 30 17:17:11 sun charon-systemd[394617]: sending packet: from 192.168.1.1[500] to 192.168.1.2[500] (592 bytes)
May 30 17:17:35 sun charon-systemd[394617]: retransmit 4 of request with message ID 0
May 30 17:17:35 sun charon-systemd[394617]: sending packet: from 192.168.1.1[500] to 192.168.1.2[500] (592 bytes)
May 30 17:18:17 sun charon-systemd[394617]: retransmit 5 of request with message ID 0
May 30 17:18:17 sun charon-systemd[394617]: sending packet: from 192.168.1.1[500] to 192.168.1.2[500] (592 bytes)
May 30 17:19:32 sun charon-systemd[394617]: creating delete job for CHILD_SA ESP/0x00000000/192.168.1.2
May 30 17:19:32 sun charon-systemd[394617]: CHILD_SA ESP/0x00000000/192.168.1.2 not found for delete
May 30 17:19:32 sun charon-systemd[394617]: giving up after 5 retransmits_

As per my understanding, both peers shall try both gateway to establish respective SA as the traffic is matching for both SA.  Now I restart the service, "gw-gw1" SA become alive and traffic start flowing in. It should have happen automatically.

**Reply by tobiasbrunner** [reply]
I already explained that before (read my first reply again).
